midp/midp.js

```diff
@@ -759,28 +759,107 @@ var MIDP = (function() {
     });
   };
 
-  var nativeEventQueues = {};
-  var waitingNativeEventQueue = {};
+  var NativeEvents = (function() {
+    var queues = {};
+    var waiting = {};
+    var penDrag = {};
+    var gestureDrag = {};
 
-  function copyEvent(e, obj) {
-    var keys = Object.keys(e);
-    for (var i = 0; i < keys.length; i++) {
-      obj[keys[i]] = e[keys[i]];
+    function copyEvent(e, obj) {
+      var keys = Object.keys(e);
+      for (var i = 0; i < keys.length; i++) {
+        obj[keys[i]] = e[keys[i]];
+      }
     }
-  }
 
-  function sendNativeEvent(e, isolateId) {
-    var elem = waitingNativeEventQueue[isolateId];
-    if (!elem) {
-      nativeEventQueues[isolateId].push(e);
-      return;
+    function reset(id) {
+      queues[id] = [];
+      delete penDrag[id];
+      delete gestureDrag[id];
+      // Don't delete waiting[id]
     }
 
-    copyEvent(e, elem.nativeEvent);
-    elem.resolve(nativeEventQueues[isolateId].length);
+    // NB: It is the responsibility of the caller to check that
+    // there is indeed an event in the queue
+    function get(id, e) {
+      var ev = queues[id].shift();
+      copyEvent(ev, e);
 
-    delete waitingNativeEventQueue[isolateId];
-  }
+      if (ev === penDrag[id]) {
+        delete penDrag[id];
+      }
+
+      if (ev === gestureDrag[id]) {
+        delete gestureDrag[id];
+      }
+    }
+
+    function numEvents(id) {
+      var q = queues[id];
+
+      if (!q) {
+        return 0;
+      }
+
+      return q.length;
+    }
+
+    // NB: It is the caller's responsibility to check that the queue is
+    // actually empty first, and to get an event instead of calling this
+    // function otherwise
+    function wait(id, resolve, e) {
+      waiting[id] = { resolve: resolve, e: e };
+    }
+
+    function put(e, id) {
+      var waiter = waiting[id];
+      if (waiter) {
+        copyEvent(e, waiter.e);
+        waiter.resolve(0);
```

Comment:
The old implementation did: `.resolve(nativeEventQueues[isolateId].length);`, but now that I think more about it you're probably right.
This will be called as soon as an event is sent to an empty queue.